aip/auth/4122.md

```diff
@@ -0,0 +1,349 @@
+# Service Account Impersonation
+
+A service account or a user identity in GCP can impersonate another service 
+account, and request short-lived credentials on behalf of that service account.
+
+The request can be direct or delegated.
+
+In the direct case, the caller is authenticated as either a Google account or a
+service account, and makes a direct request to create short-lived credentials on
+behalf of another service account. The requesting identity must have the Service
+Account Token Creator role (`roles/iam.serviceAccountTokenCreator`) on the target
+service account.
+
+In the delegated case, the caller is also authenticated as either a Google
+account or a service account, but the request goes through one or more service
+accounts in a delegation chain. Each identity in the chain must grant the Service
+Account Token Creator role to the previous identity in the chain. 
+
+**Note:** Because this AIP describes guidance and requirements in a
+language-neutral way, it uses generic terminology which may be imprecise or
+inappropriate in certain languages or environments.
+
+## Guidance
+
+This section describes the general guidance for supporting Service Account
+Impersonation from an auth library to generate OAuth 2.0 access tokens,
+OpenID Connect ID tokens, self-signed JSON Web Tokens and self-signed blobs
+ on behalf of an impersonated service account.
+
+For instructions on how to configure the correct permissions for impersonation,
+please read the [Creating short-lived service account credentials][0] docs.
+
+For general guidance on supporting Service Accounts with Service Account Keys,
+please read the [Service Account Keys][1] AIP.
+
+### Generating an OAuth 2.0 access token
+
+#### Parameters
+
+The auth library **must** implement the language-specific abstractions required
+to allow a caller to specify the following parameters for a request to generate
+an OAuth 2.0 access token on behalf of a service account.
+
+**Source Credentials** - **Required**
+
+The library specific representation of the service/user account requesting
+ the impersonation.
+ 
+**Target Principal** - **Required**
+
+The email address of the service account to impersonate, e.g. 
+`sa-name@project.iam.gserviceaccount.com`.
+
+**Scopes** - **Required**
+
+A list of requested OAuth2.0 scopes.
+
+Auth libraries **must** validate that the list of requested scopes is not empty,
+and throw an error in that case to save a round trip for a request that would
+fail.
+
+**Lifetime** - **Required**
+
+The amount of time until the access token expires, in seconds.
+By default, the maximum lifetime for a token is 1 hour (3,600 seconds) and the
+value can be extended to 12 hours (43,200 seconds) adding the target service
+account to an organization policy that includes
+`constraints/iam.allowServiceAccountCredentialLifetimeExtension`.
+
+Auth libraries **must** validate that lifetime doesn't exceed the potential
+maximum of 43,200 seconds, and throw an error if it does.
```

Comment:
I think we should rely on the server to validate this, instead of relying on a hardcoded number in clients, as the maximum lifetime can be changed in the future.